File: ros2/src/bebop_pilot/bebop_pilot/robstride.py
"""Robstride RS01/RS02/RS03/RS04 CAN protocol — pure Python.

Ported from the Rust ``firmware/bebop-linux/src/{robstride,can_interface}.rs``.

No ROS dependencies; only ``struct`` and ``math`` from the standard library.
The CAN transport (python-can / socketcan) is injected as plain (id, data)
tuples so this module is unit-testable without hardware.

Terminology:
    motor_id          — 8-bit Robstride node ID, set in the Robstride
                        debugger per motor (e.g. 31, 41, 32, ...).
    frame_id          — full 29-bit extended CAN identifier sent on the
                        wire, built from cmd_type + data_area2 + motor_id.

Wire format (Robstride MIT-mode, CAN 2.0 Extended, 1 Mbps):

    29-bit frame_id (TX, host -> motor)
        bits 28..24 : cmd_type        (5 bits)
        bits 23.. 8 : data area 2     (16 bits, e.g. torque feedforward)
        bits  7.. 0 : motor_id        (8 bits)

    8-byte payload for cmd_type = MOTOR_CTRL (0x01), big-endian:
        [0..2)  position_raw   (uint16, -4pi..+4pi)
        [2..4)  velocity_raw   (uint16, model-specific)
        [4..6)  kp_raw         (uint16, 0..5000)
        [6..8)  kd_raw         (uint16, 0..100)

    29-bit frame_id (RX, motor -> host, cmd_type = 0x02)
        bits 28..24 : cmd_type = 0x02
        bits 23..22 : mode_status
        bits 21..16 : fault_bits
        bits 15.. 8 : motor_id  (which motor responded)
        bits  7.. 0 : host_id   (0xFD)

    8-byte payload, big-endian:
        [0..2)  position_raw   (uint16, -4pi..+4pi)
        [2..4)  velocity_raw   (uint16, model-specific)
        [4..6)  torque_raw     (uint16, model-specific)
        [6..8)  temperature_raw (uint16, °C * 10)
"""

from __future__ import annotations

import math
import struct
from dataclasses import dataclass
from enum import IntEnum
from typing import Dict, Optional, Tuple


# ----------------------------------------------------------------- protocol
HOST_ID: int = 0xFD

P_MIN: float = -4.0 * math.pi  # -12.566...
P_MAX: float = +4.0 * math.pi
KP_MIN: float = 0.0
KP_MAX: float = 5000.0
KD_MIN: float = 0.0
KD_MAX: float = 100.0


class CmdType(IntEnum):
    GET_ID = 0x00
    MOTOR_CTRL = 0x01
    FEEDBACK = 0x02
    ENABLE = 0x03
    STOP = 0x04
    SET_ZERO = 0x06
    PARAM_READ = 0x11
    PARAM_WRITE = 0x12
    FAULT_FEEDBACK = 0x15
    ACTIVE_REPORT = 0x18


# ----------------------------------------------------------------- model specs
@dataclass(frozen=True)
class RobstrideSpecs:
    """Per-model torque/velocity ranges used for uint16 <-> float scaling."""
    name: str
    torque_min: float
    torque_max: float
    velocity_min: float
    velocity_max: float


SPECS: Dict[str, RobstrideSpecs] = {
    "RS01": RobstrideSpecs("RS01", -12.0, 12.0, -45.0, 45.0),
    "RS02": RobstrideSpecs("RS02", -25.0, 25.0, -30.0, 30.0),
    "RS03": RobstrideSpecs("RS03", -60.0, 60.0, -20.0, 20.0),
    "RS04": RobstrideSpecs("RS04", -120.0, 120.0, -15.0, 15.0),
}


def get_specs(model: str) -> RobstrideSpecs:
    try:
        return SPECS[model.upper()]
    except KeyError as exc:
        raise ValueError(
            f"Unknown Robstride model {model!r}. Expected one of {list(SPECS)}."
        ) from exc


# ----------------------------------------------------------------- helpers
def _f_to_u16(value: float, lo: float, hi: float) -> int:
    if hi <= lo:
        raise ValueError(f"Invalid range [{lo}, {hi}]")
    clamped = min(max(value, lo), hi)
    return int(round((clamped - lo) / (hi - lo) * 65535.0))


def _u16_to_f(raw: int, lo: float, hi: float) -> float:
    return lo + (raw / 65535.0) * (hi - lo)


# ----------------------------------------------------------------- frame builders
def _make_frame_id(cmd_type: int, data_area2: int, motor_id: int) -> int:
    if not (0 <= cmd_type <= 0x1F):
        raise ValueError(f"cmd_type out of range: {cmd_type:#x}")
    if not (0 <= data_area2 <= 0xFFFF):
        raise ValueError(f"data_area2 out of range: {data_area2:#x}")
    if not (0 <= motor_id <= 0xFF):
        raise ValueError(f"motor_id out of range: {motor_id:#x}")
    return ((cmd_type & 0x1F) << 24) | ((data_area2 & 0xFFFF) << 8) | (motor_id & 0xFF)


def build_enable(motor_id: int) -> Tuple[int, bytes]:
    """Frame to enable a motor (transition to operation mode)."""
    return _make_frame_id(CmdType.ENABLE, HOST_ID, motor_id), bytes(8)


def build_disable(motor_id: int) -> Tuple[int, bytes]:
    """Frame to stop a motor (transition out of operation mode)."""
    return _make_frame_id(CmdType.STOP, HOST_ID, motor_id), bytes(8)


def build_set_zero(motor_id: int) -> Tuple[int, bytes]:
    """Frame to set the current position as the new mechanical zero."""
    payload = bytes([1, 0, 0, 0, 0, 0, 0, 0])
    return _make_frame_id(CmdType.SET_ZERO, HOST_ID, motor_id), payload


def build_active_report(motor_id: int, interval_ms: int = 5) -> Tuple[int, bytes]:
    """Ask the motor to stream feedback every ``interval_ms`` ms.

    Default 5 ms = 200 Hz, well above the 100 Hz training control loop.
    """
    if not (1 <= interval_ms <= 255):
        raise ValueError("interval_ms must be in [1, 255]")
    payload = bytes([1, interval_ms, 0, 0, 0, 0, 0, 0])
    return _make_frame_id(CmdType.ACTIVE_REPORT, HOST_ID, motor_id), payload


def build_motor_ctrl(
    motor_id: int,
    model: str,
    position: float,
    velocity: float = 0.0,
    kp: float = 0.0,
    kd: float = 0.0,
    torque_ff: float = 0.0,
) -> Tuple[int, bytes]:
    """Build a MIT-mode control frame for the motor with ``motor_id``.

    ``torque_ff`` is the feed-forward torque (Nm) packed into the ID's data
    area 2 — it's *not* in the payload. The payload carries the PD setpoint.

    Output: (frame_id, 8-byte payload), where ``frame_id`` is the full
    29-bit extended CAN identifier. Big-endian throughout.
    """
    specs = get_specs(model)

    pos_raw = _f_to_u16(position, P_MIN, P_MAX)
    vel_raw = _f_to_u16(velocity, specs.velocity_min, specs.velocity_max)
    kp_raw = _f_to_u16(kp, KP_MIN, KP_MAX)
    kd_raw = _f_to_u16(kd, KD_MIN, KD_MAX)
    tau_raw = _f_to_u16(torque_ff, specs.torque_min, specs.torque_max)

    frame_id = _make_frame_id(CmdType.MOTOR_CTRL, tau_raw, motor_id)
    payload = struct.pack(">HHHH", pos_raw, vel_raw, kp_raw, kd_raw)
    return frame_id, payload


# ----------------------------------------------------------------- feedback parser
@dataclass
class Feedback:
    """Parsed Robstride feedback frame."""
    motor_id: int
    host_id: int
    fault_bits: int
    mode_status: int  # 0=reset, 1=cali, 2=motor (enabled), ...
    position: float   # rad
    velocity: float   # rad/s
    torque: float     # Nm
    temperature: float  # °C

    @property
    def is_enabled(self) -> bool:
        return self.mode_status == 2

    @property
    def has_fault(self) -> bool:
        return self.fault_bits != 0

    def fault_description(self) -> str:
        if self.fault_bits == 0:
            return ""
        flags = []
        if self.fault_bits & 0x01:
            flags.append("undervoltage")
        if self.fault_bits & 0x02:
            flags.append("overcurrent")
        if self.fault_bits & 0x04:
            flags.append("overtemperature")
        if self.fault_bits & 0x08:
            flags.append("encoder_fault")
        if self.fault_bits & 0x10:
            flags.append("gridlock_overload")
        if self.fault_bits & 0x20:
            flags.append("uncalibrated")
        return ",".join(flags) if flags else f"unknown(0x{self.fault_bits:02x})"


def parse_feedback(frame_id: int, data: bytes, model: str) -> Optional[Feedback]:
    """Parse an extended-frame feedback message. Returns None if not feedback.

    ``frame_id`` is the full 29-bit extended CAN identifier as received from
    the bus. The model is needed for the velocity / torque uint16 ranges
    since the Robstride frame does not encode the model.
    """
    if len(data) < 8:
        return None

    cmd_type = (frame_id >> 24) & 0x1F
    if cmd_type != CmdType.FEEDBACK:
        return None

    mode_status = (frame_id >> 22) & 0x03
    fault_bits = (frame_id >> 16) & 0x3F
    motor_id = (frame_id >> 8) & 0xFF
    host_id = frame_id & 0xFF

    pos_raw, vel_raw, tau_raw, temp_raw = struct.unpack(">HHHH", data[:8])

    specs = get_specs(model)
    position = _u16_to_f(pos_raw, P_MIN, P_MAX)
    velocity = _u16_to_f(vel_raw, specs.velocity_min, specs.velocity_max)
    torque = _u16_to_f(tau_raw, specs.torque_min, specs.torque_max)
    temperature = temp_raw / 10.0

    return Feedback(
        motor_id=motor_id,
        host_id=host_id,
        fault_bits=fault_bits,
        mode_status=mode_status,
        position=position,
        velocity=velocity,
        torque=torque,
        temperature=temperature,
    )

File: firmware/bebop-linux/src/robstride.rs
//! Robstride motor driver (RS01, RS02, RS03, RS04)
//!
//! Implements the Robstride CAN protocol for motor control.
//! Uses CAN 2.0 Extended Frame (29-bit ID) at 1 Mbps.

use crate::can_interface::{CanInterface, ReceivedFrame, RobstrideFeedback};
use crate::config::{JointCommand, JointState, RobstrideModel, RobstrideSpecs};
use anyhow::Result;
use tracing::{debug, trace};

/// Robstride communication types (bits 28-24 of extended frame ID)
mod cmd {
    pub const GET_ID: u8 = 0x00;
    pub const MOTOR_CTRL: u8 = 0x01;    // Operation control mode
    pub const FEEDBACK: u8 = 0x02;       // Motor feedback response
    pub const ENABLE: u8 = 0x03;
    pub const STOP: u8 = 0x04;
    pub const SET_ZERO: u8 = 0x06;
    pub const PARAM_READ: u8 = 0x11;
    pub const PARAM_WRITE: u8 = 0x12;
    pub const FAULT_FEEDBACK: u8 = 0x15;
    pub const ACTIVE_REPORT: u8 = 0x18;
}

/// Robstride parameter indices
mod param {
    pub const RUN_MODE: u16 = 0x7005;
    pub const IQ_REF: u16 = 0x7006;
    pub const SPD_REF: u16 = 0x700A;
    pub const LIMIT_TORQUE: u16 = 0x700B;
    pub const LOC_REF: u16 = 0x7016;
    pub const LIMIT_SPD: u16 = 0x7017;
    pub const LIMIT_CUR: u16 = 0x7018;
    pub const MECH_POS: u16 = 0x7019;
    pub const IQF: u16 = 0x701A;
    pub const MECH_VEL: u16 = 0x701B;
    pub const VBUS: u16 = 0x701C;
}

/// Run modes
mod mode {
    pub const OPERATION: u8 = 0x00;      // MIT-like control
    pub const POSITION_PP: u8 = 0x01;    // Profile Position
    pub const VELOCITY: u8 = 0x02;
    pub const CURRENT: u8 = 0x03;
    pub const POSITION_CSP: u8 = 0x05;   // Cyclic Synchronous Position
}

/// Protocol constants (same for all models)
const P_MIN: f32 = -12.57;  // -4π rad
const P_MAX: f32 = 12.57;   // +4π rad
const KP_MIN: f32 = 0.0;
const KP_MAX: f32 = 5000.0;
const KD_MIN: f32 = 0.0;
const KD_MAX: f32 = 100.0;

/// Host CAN ID (master controller)
const HOST_ID: u8 = 0xFD;

/// Robstride motor driver
#[derive(Debug)]
pub struct RobstrideMotor {
    pub can_id: u8,
    pub model: RobstrideModel,
    specs: RobstrideSpecs,
    pub state: JointState,
    run_mode: u8,
}

impl RobstrideMotor {
    /// Create a new Robstride motor driver
    pub fn new(can_id: u8, model: RobstrideModel) -> Self {
        Self {
            can_id,
            model,
            specs: model.specs(),
            state: JointState::default(),
            run_mode: mode::OPERATION,
        }
    }

    /// Build the 29-bit extended CAN ID
    fn make_can_id(&self, cmd_type: u8, data_area2: u16) -> u32 {
        let id = ((cmd_type as u32) << 24)
            | ((data_area2 as u32) << 8)
            | (self.can_id as u32);
        id
    }

    /// Build CAN ID for operation control mode
    fn make_ctrl_can_id(&self, torque_raw: u16) -> u32 {
        ((cmd::MOTOR_CTRL as u32) << 24)
            | ((torque_raw as u32) << 8)
            | (self.can_id as u32)
    }

    /// Convert float to uint16 with given range
    fn float_to_uint16(value: f32, min: f32, max: f32) -> u16 {
        let clamped = value.clamp(min, max);
        let proportion = (clamped - min) / (max - min);
        (proportion * 65535.0) as u16
    }

    /// Convert uint16 to float with given range
    fn uint16_to_float(value: u16, min: f32, max: f32) -> f32 {
        let proportion = value as f32 / 65535.0;
        min + proportion * (max - min)
    }

    /// Enable the motor
    pub fn enable(&self, can: &CanInterface) -> Result<()> {
        let can_id = self.make_can_id(cmd::ENABLE, HOST_ID as u16);
        let data = [0u8; 8];
        can.send_extended(can_id, &data)?;
        debug!("Enabled Robstride motor {}", self.can_id);
        Ok(())
    }

    /// Disable the motor
    pub fn disable(&self, can: &CanInterface) -> Result<()> {
        let can_id = self.make_can_id(cmd::STOP, HOST_ID as u16);
        let data = [0u8; 8];
        can.send_extended(can_id, &data)?;
        debug!("Disabled Robstride motor {}", self.can_id);
        Ok(())
    }

    /// Set mechanical zero position
    pub fn set_zero(&self, can: &CanInterface) -> Result<()> {
        let can_id = self.make_can_id(cmd::SET_ZERO, HOST_ID as u16);
        let mut data = [0u8; 8];
        data[0] = 1;
        can.send_extended(can_id, &data)?;
        debug!("Set zero position for Robstride motor {}", self.can_id);
        Ok(())
    }

    /// Enable active reporting at specified interval (ms)
    pub fn enable_active_reporting(&self, can: &CanInterface, interval_ms: u8) -> Result<()> {
        let can_id = self.make_can_id(cmd::ACTIVE_REPORT, HOST_ID as u16);
        let mut data = [0u8; 8];
        data[0] = 1;  // Enable
        data[1] = interval_ms;
        can.send_extended(can_id, &data)?;
        debug!(
            "Enabled active reporting for Robstride motor {} at {}ms",
            self.can_id, interval_ms
        );
        Ok(())
    }

    /// Send operation control mode command (MIT-like control)
    ///
    /// τ = Kp × (p_target - p_actual) + Kd × (v_target - v_actual) + τ_ff
    pub fn send_command(&self, can: &CanInterface, cmd: &JointCommand) -> Result<()> {
        // Convert to raw values
        let position_raw = Self::float_to_uint16(cmd.position, P_MIN, P_MAX);
        let velocity_raw = Self::float_to_uint16(
            cmd.velocity,
            self.specs.velocity_min,
            self.specs.velocity_max,
        );
        let kp_raw = Self::float_to_uint16(cmd.kp, KP_MIN, KP_MAX);
        let kd_raw = Self::float_to_uint16(cmd.kd, KD_MIN, KD_MAX);
        let torque_raw = Self::float_to_uint16(
            cmd.torque,
            self.specs.torque_min,
            self.specs.torque_max,
        );

        // Build CAN frame
        let can_id = self.make_ctrl_can_id(torque_raw);
        let mut data = [0u8; 8];

        // Pack data (big-endian)
        data[0..2].copy_from_slice(&position_raw.to_be_bytes());
        data[2..4].copy_from_slice(&velocity_raw.to_be_bytes());
        data[4..6].copy_from_slice(&kp_raw.to_be_bytes());
        data[6..8].copy_from_slice(&kd_raw.to_be_bytes());

        can.send_extended(can_id, &data)?;

        trace!(
            "RS{} CMD: pos={:.3} vel={:.3} kp={:.1} kd={:.2} τ={:.2}",
            self.can_id,
            cmd.position,
            cmd.velocity,
            cmd.kp,
            cmd.kd,
            cmd.torque
        );

        Ok(())
    }

    /// Send position-only command (uses configured Kp/Kd)
    pub fn send_position(&self, can: &CanInterface, position: f32, kp: f32, kd: f32) -> Result<()> {
        let cmd = JointCommand {
            position,
            velocity: 0.0,
            torque: 0.0,
            kp,
            kd,
        };
        self.send_command(can, &cmd)
    }

    /// Process feedback frame
    pub fn process_feedback(&mut self, feedback: &RobstrideFeedback) {
        if feedback.motor_id != self.can_id {
            return;
        }

        self.state.position = feedback.position;
        self.state.velocity = feedback.velocity;
        self.state.torque = feedback.torque;
        self.state.temperature = feedback.temperature;
        self.state.has_error = feedback.fault_bits != 0;
        self.state.error_code = feedback.fault_bits as u32;
        self.state.is_enabled = feedback.mode_status == 0x02;  // Motor mode
        self.state.last_update_ms = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_millis() as u64;
    }

    /// Check if motor is receiving feedback
    pub fn is_alive(&self, max_age_ms: u64) -> bool {
        let now = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_millis() as u64;
        
        (now - self.state.last_update_ms) < max_age_ms
    }

    /// Get fault description
    pub fn fault_description(&self) -> Option<String> {
        let bits = self.state.error_code as u8;
        if bits == 0 {
            return None;
        }

        let mut faults = Vec::new();
        if bits & 0x01 != 0 { faults.push("Undervoltage"); }
        if bits & 0x02 != 0 { faults.push("Overcurrent"); }
        if bits & 0x04 != 0 { faults.push("Overtemperature"); }
        if bits & 0x08 != 0 { faults.push("Magnetic encoding fault"); }
        if bits & 0x10 != 0 { faults.push("Gridlock overload"); }
        if bits & 0x20 != 0 { faults.push("Uncalibrated"); }

        Some(faults.join(", "))
    }
}

/// Collection of Robstride motors on a single bus
pub struct RobstrideMotorBus {
    pub motors: Vec<RobstrideMotor>,
}

impl RobstrideMotorBus {
    pub fn new() -> Self {
        Self { motors: Vec::new() }
    }

    pub fn add_motor(&mut self, can_id: u8, model: RobstrideModel) {
        self.motors.push(RobstrideMotor::new(can_id, model));
    }

    pub fn get_motor(&self, can_id: u8) -> Option<&RobstrideMotor> {
        self.motors.iter().find(|m| m.can_id == can_id)
    }

    pub fn get_motor_mut(&mut self, can_id: u8) -> Option<&mut RobstrideMotor> {
        self.motors.iter_mut().find(|m| m.can_id == can_id)
    }

    /// Enable all motors
    pub fn enable_all(&self, can: &CanInterface) -> Result<()> {
        for motor in &self.motors {
            motor.enable(can)?;
            std::thread::sleep(std::time::Duration::from_millis(10));
        }
        Ok(())
    }

    /// Disable all motors
    pub fn disable_all(&self, can: &CanInterface) -> Result<()> {
        for motor in &self.motors {
            motor.disable(can)?;
            std::thread::sleep(std::time::Duration::from_millis(10));
        }
        Ok(())
    }

    /// Process received frame and update motor state
    pub fn process_frame(&mut self, frame: &ReceivedFrame) {
        if let Some(feedback) = frame.parse_robstride() {
            if let Some(motor) = self.get_motor_mut(feedback.motor_id) {
                motor.process_feedback(&feedback);
            }
        }
    }
}

impl Default for RobstrideMotorBus {
    fn default() -> Self {
        Self::new()
    }
}
